prov/gni/src/gnix_vc.c

```diff
@@ -1661,27 +1656,28 @@ static int __gnix_vc_push_work_reqs(struct gnix_vc *vc)
 				  "Request processed: %p\n", req);
 			goto try_again;
 		} else {
-			/* Work failed.  Reschedule to put this VC back on the
-			 * end of the list. */
+			/* Work failed.  Reschedule to put this VC
+			 * back on the end of the list and return
+			 * -FI_EAGAIN */
 			__gnix_vc_work_schedule(vc);
 
 			fastlock_acquire(&vc->work_queue_lock);
 			slist_insert_tail(item, &vc->work_queue);
 			fastlock_release(&vc->work_queue_lock);
 
-			/* FI_ENOSPC is reserved to indicate a lack of TXDs,
-			 * which are shared by all VCs on the NIC.  Return
-			 * error to stall processing of VCs in this case.  The
-			 * other likely error is a lack of SMSG credits, which
-			 * only halts this VC. */
-			if (ret == -FI_ENOSPC) {
-				fi_rc = -FI_EAGAIN;
-			} else if (ret != -FI_EAGAIN) {
+			fi_rc = -FI_EAGAIN;
+
+			/* FI_ENOSPC is reserved to indicate a lack of
+			 * TXDs, which are shared by all VCs on the
+			 * NIC.  The other likely error is FI_EAGAIN
+			 * due to a lack of SMSG credits. */
+
+			if ((ret != -FI_ENOSPC) && (ret != -FI_EAGAIN)) {
 				/* TODO report error? */
 				GNIX_WARN(FI_LOG_EP_DATA,
 					  "Failed to push request %p: %s\n",
 					  req, fi_strerror(-ret));
-			} /* else return success to keep processing TX VCs */
+			}
 		}
 
 		goto try_again;
```

Comment:
Are we sure this loop isn't a problem?  I think we were supposed to break out if fi_rc was set.  At some point it was converted to use goto and it doesn't look like we break out in the failure case anymore.